algosdk/future/atomic_transaction_composer.py

```diff
@@ -0,0 +1,373 @@
+            return self.signed_txns
+
+        txn_list = self.build_group()
+        for txn_with_signer in txn_list:
+            unsigned_txn = txn_with_signer.txn
+            stxn = txn_with_signer.signer.sign_txn(unsigned_txn)
+            self.signed_txns.append(stxn)
+
+        self.status = AtomicTransactionComposerStatus.SIGNED
+        return self.signed_txns
+
+    def submit(self, client):
+        """
+        Send the transaction group to the network, but don't wait for it to be
+        committed to a block. An error will be thrown if submission fails.
+        The composer's status must be SUBMITTED or lower before calling this method.
+        If submission is successful, this composer's status will update to SUBMITTED.
+
+        Note: a group can only be submitted again if it fails.
+
+        Returns:
+            list[Transaction]: list of submitted transactions
+        """
+        if self.status <= AtomicTransactionComposerStatus.SUBMITTED:
+            self.build_group()
+            self.gather_signatures()
+        else:
+            raise error.AtomicTransactionComposerError(
+                "AtomicTransactionComposerStatus must be submitted or lower to submit a group"
+            )
+
+        client.send_transactions(self.signed_txns)
+        self.status = AtomicTransactionComposerStatus.SUBMITTED
+        return self.tx_ids
+
+    def execute(self, client):
+        """
+        Send the transaction group to the network and wait until it's committed
+        to a block. An error will be thrown if submission or execution fails.
+        The composer's status must be SUBMITTED or lower before calling this method,
+        since execution is only allowed once. If submission is successful,
+        this composer's status will update to SUBMITTED.
+        If the execution is also successful, this composer's status will update to COMMITTED.
+
+        Note: a group can only be submitted again if it fails.
+
+        Returns:
+            AtomicTransactionResponse: Object with confirmed round for this transaction,
+                a list of txIDs of the submitted transactions, and an array of
+                results for each method call transaction in this group. If a
+                method has no return value (void), then the method results array
+                will contain None for that method's return value.
+        """
+        if self.status > AtomicTransactionComposerStatus.SUBMITTED:
+            raise error.AtomicTransactionComposerError(
+                "AtomicTransactionComposerStatus must be submitted or lower to execute a group"
+            )
+
+        client.submit(client)
+        client.wait_for_confirmation(client, self.tx_ids[0])
+
+        resp = client.pending_transaction_info(self.tx_ids[0])
+        resp_dict = json.loads(resp)
+
+        confirmed_round = resp_dict["confirmed-round"]
+        tx_ids = self.tx_ids
+        results = resp_dict["logs"] if "logs" in resp_dict else None
```

Comment:
From Jason: "We have not finalized this 100% yet, but the plan is to have the contract log the return value as encoded bytes, likely with a special 4-byte prefix (maybe all 0s). So you can use the pending txn endpoint to see the logged values, then iterate over them to find an entry that starts with 0x00000000, and decode the rest of it to get the ABI return value"

Reply:
Thanks! Wasn't aware of this -- I'll change this